solidity/contracts/libraries/operator/GroupSelection.sol

```diff
@@ -404,4 +418,21 @@ library GroupSelection {
             delete self.candidate[self.tickets[i]];
         }
     }
+
+    function reactiveTicketSubmissionRounds(
+        uint256 possibleOperatorsCount,
+        uint256 groupSize
+    ) public view returns (uint256) {
+        uint256 operatorsExponent = 1;
+        while(2**operatorsExponent < possibleOperatorsCount){
+            operatorsExponent++;
+        }
+
+        uint256 groupsExponent = 1;
+        while(2**groupsExponent < groupSize){
+            groupsExponent++;
+        }
+
+        return operatorsExponent.sub(groupsExponent);
+    }
```

Comment:
The maximum number of virtual stakers after the launch is `1e9 / 1e5 = 1e4`. The operator exponent is `2^13`. The group exponent is `2^5` so the reactive ticket submission rounds count is `2^13 - 2^5 = 8160`. Assuming a block time of 10 seconds, it's about 2 hours.

The maximum number of virtual stakers after 2 years is `1e9 / 1e4 = 1e5`. The operator exponent is `2^16`. The group exponent is `2^5` so the reactive ticket submission rounds count is `2^16 - 2^5 = 65504`. Assuming a block time of 10 seconds, it's about 18 hours.

Is this correct? 

I'd greatly appreciate @eth-r look on the on-chain part here.

Another question - do we need to have all those calculations here? Can't we hardcode the number of rounds to `10k` minimum stake case? This will simplify the on-chain code and we are probably not losing that much given that we'll eventually be at that point.

Reply:
The round count is logarithms so it's `13 - 5 = 8`. We should hardcode the values in a constant; if we upgrade the beacon operator contract we can compile in new values, so there is no real need for them to be storage.

Reply:
Yeah, they should be logarithms but I don't see logarithms in the code, that's why I was so confused. 

`13 - 5 = 8` is the launch case but `16 - 5 = 11` is the two-years case. It's just `3` rounds more. Let's just have a constant number of reactive rounds of `11` (plus initial phase and mining lag at the end) and if we don't use all rounds for a post-launch case, we'll treat `11-8=3` as an additional mining lag buffer at the end. 

What do you think?

Reply:
The logarithm is in: `while(2**groupsExponent < groupSize)`

Every time `groupsExponent` is incremented, the value of `2**groupsExponent` doubles.

And yes, we should have a constant 11 rounds plus the mining lag. We don't need the initial phase separately from the zero round.

Reply:
> The logarithm is in: `while(2**groupsExponent < groupSize)`
> 
> Every time `groupsExponent` is incremented, the value of `2**groupsExponent` doubles.

Ah, of course.

> And yes, we should have a constant 11 rounds plus the mining lag. We don't need the initial phase separately from the zero round.

Not sure if I follow this one. Don't we need the `< natural_threshold` round?

Reply:
The zero round is the natural threshold round; operators with 11 or more leading zeros all submit in the zeroth round.

Reply:
Synced with @eth-r out-of-band. The summary is:

We have `v=11` rounds. It is a constant, no need to calculate it in the contract. We start indexing rounds from `0` and we do:
- `round 0`: submit tickets with `11` or more leading zeros
- `round 1`: submit tickets with `10` leading zeros
- `round 2`: submit tickets with `9` leading zeros
...
- round `11`: submit tickets with `0` leading zeros

(we submit in the given round only if it still makes sense)

We no longer need the initial submission phase and we no longer calculate the natural threshold. We submit tickets based on the leading zeros and that's all. 

Round 11 is different from the other rounds; its duration has to be the entire mining lag even if other rounds are shorter. We have 11 rounds of `roundDuration` and then a post-round of `miningLag`. `miningLag = 6, roundDuration = 3`

Reply:
Refactored the code according to the above comments. But, regarding the on-chain part, I've just increased the submission timeout constant and keep the contract not aware of rounds. I've done that because the contract doesn't need to know about them, so we can avoid problems with contract size if we add some additional fields (even if we put new fields in `GroupSelection` library we must add some getters in operator contract).

Reply:
Sounds like a good move given that it's the client who is interested in the optimization, not the chain.